code/game/objects/items/jetpack.dm

```diff
@@ -0,0 +1,173 @@
+	icon = 'icons/obj/items/jetpack.dmi'
+	icon_state = "jetpack_marine"
+	w_class = WEIGHT_CLASS_BULKY
+	flags_equip_slot = ITEM_SLOT_BACK
+	///maximum amount of fuel in the jetpack
+	var/fuel_max = 60
+	///current amount of fuel in the jetpack
+	var/fuel_left = 60
+	///threshold to change the jetpack fuel indicator
+	var/fuel_indicator = 40
+	///How quick you will fly (warning, it rounds up to the nearest integer)
+	var/speed = 1
+	///How long the jetpack allows you to fly over things
+	var/hovering_time = 1 SECONDS
+	///True when jetpack has flame overlay
+	var/lit = FALSE
+	///True if you can use shift click/middle click to use it
+	var/selected = FALSE
+
+/obj/item/jetpack_marine/Initialize()
+	. = ..()
+	update_icon()
+
+/obj/item/jetpack_marine/examine(mob/user, distance, infix, suffix)
+	. = ..()
+	if(!ishuman(user))
+		return
+	if(fuel_left == 0)
+		to_chat(user, "The fuel gauge beeps out, it has no fuel left")
+	else
+		to_chat(user, "The fuel gauge meter indicate it has [fuel_left/FUEL_USE] uses left")
+
+/obj/item/jetpack_marine/equipped(mob/user, slot)
+	. = ..()
+	if(slot == SLOT_BACK)
+		RegisterSignal(user, COMSIG_MOB_CLICK_ALT_RIGHT, .proc/try_to_use_jetpack)
+		var/datum/action/item_action/toggle/action = new(src)
+		action.give_action(user)
+
+/obj/item/jetpack_marine/dropped(mob/user)
+	. = ..()
+	UnregisterSignal(user, list(COMSIG_MOB_CLICK_ALT_RIGHT, COMSIG_MOB_MIDDLE_CLICK))
+	UnregisterSignal(user, COMSIG_ITEM_EXCLUSIVE_TOGGLE)
+	selected = FALSE
+	actions.Cut()
+
+/obj/item/jetpack_marine/ui_action_click(mob/user, datum/action/item_action/action)
+	if(selected)
+		UnregisterSignal(user, COMSIG_MOB_MIDDLE_CLICK)
+		action.remove_selected_frame()
+		UnregisterSignal(user, COMSIG_ITEM_EXCLUSIVE_TOGGLE)
+	else
+		RegisterSignal(user, COMSIG_MOB_MIDDLE_CLICK, .proc/try_to_use_jetpack)
+		action.add_selected_frame()
+		SEND_SIGNAL(user, COMSIG_ITEM_EXCLUSIVE_TOGGLE, user)
+		RegisterSignal(user, COMSIG_ITEM_EXCLUSIVE_TOGGLE, .proc/unselect)
+	selected = !selected
+
+///Signal handler for making it impossible to use middleclick to use the jetpack
+/obj/item/jetpack_marine/proc/unselect(datum/source, mob/user)
+	SIGNAL_HANDLER
+	if(!selected)
+		return
+	selected = FALSE
+	UnregisterSignal(src, COMSIG_ITEM_MIDDLECLICKON)
+	UnregisterSignal(user, COMSIG_ITEM_EXCLUSIVE_TOGGLE)
+	for(var/datum/action/action in user.actions)
+		if (!istype(action, /datum/action/item_action))
+			continue
+		var/datum/action/item_action/iaction = action
+		if(iaction?.holder_item == src)
+			action.remove_selected_frame()
+
+
+///remove the flame overlay
+/obj/item/jetpack_marine/proc/reset_flame(mob/living/carbon/human/human_user)
+	lit = FALSE
+	update_icon()
+	human_user.update_inv_back()
+
+///Signal handler for alt click, when the user want to fly at an atom
+/obj/item/jetpack_marine/proc/try_to_use_jetpack(datum/source, atom/A)
+	SIGNAL_HANDLER
+	var/mob/living/carbon/human/human_user = usr
+	if (!use_jetpack(human_user))
+		return
```

Comment:
These two procs seem very obfuscated for what the name seems to do (try_to_use just calls use_jetpack, which does the actual checks but the the actual jetpacking is done in try_to_use
this is nonblocking but it confused me when reviewing